From: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Subject: fs: consistently use file_has_valid_mmap_hooks() helper
Date: Mon, 16 Jun 2025 20:33:22 +0100

Since commit c84bf6dd2b83 ("mm: introduce new .mmap_prepare() file
callback"), the f_op->mmap() hook has been deprecated in favour of
f_op->mmap_prepare().

Additionally, commit bb666b7c2707 ("mm: add mmap_prepare() compatibility
layer for nested file systems") permits the use of the .mmap_prepare()
hook even in nested filesystems like overlayfs.

There are a number of places where we check only for f_op->mmap - this is
incorrect now mmap_prepare exists, so update all of these to use the
general helper file_has_valid_mmap_hooks().

Most notably, this updates the elf logic to allow for the ability to
execute binaries on filesystems which have the .mmap_prepare hook, but
additionally we update nested filesystems.

Link: https://lkml.kernel.org/r/b68145b609532e62bab603dd9686faa6562046ec.1750099179.git.lorenzo.stoakes@oracle.com
Signed-off-by: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 fs/backing-file.c     |    2 +-
 fs/binfmt_elf.c       |    4 ++--
 fs/binfmt_elf_fdpic.c |    2 +-
 fs/coda/file.c        |    2 +-
 fs/ecryptfs/file.c    |    2 +-
 5 files changed, 6 insertions(+), 6 deletions(-)

--- a/fs/backing-file.c~fs-consistently-use-file_has_valid_mmap_hooks-helper
+++ a/fs/backing-file.c
@@ -333,7 +333,7 @@ int backing_file_mmap(struct file *file,
 	if (WARN_ON_ONCE(!(file->f_mode & FMODE_BACKING)))
 		return -EIO;
 
-	if (!file->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(file))
 		return -ENODEV;
 
 	vma_set_file(vma, file);
--- a/fs/binfmt_elf.c~fs-consistently-use-file_has_valid_mmap_hooks-helper
+++ a/fs/binfmt_elf.c
@@ -646,7 +646,7 @@ static unsigned long load_elf_interp(str
 	if (!elf_check_arch(interp_elf_ex) ||
 	    elf_check_fdpic(interp_elf_ex))
 		goto out;
-	if (!interpreter->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(interpreter))
 		goto out;
 
 	total_size = total_mapping_size(interp_elf_phdata,
@@ -848,7 +848,7 @@ static int load_elf_binary(struct linux_
 		goto out;
 	if (elf_check_fdpic(elf_ex))
 		goto out;
-	if (!bprm->file->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(bprm->file))
 		goto out;
 
 	elf_phdata = load_elf_phdrs(elf_ex, bprm->file);
--- a/fs/binfmt_elf_fdpic.c~fs-consistently-use-file_has_valid_mmap_hooks-helper
+++ a/fs/binfmt_elf_fdpic.c
@@ -109,7 +109,7 @@ static int is_elf(struct elfhdr *hdr, st
 		return 0;
 	if (!elf_check_arch(hdr))
 		return 0;
-	if (!file->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(file))
 		return 0;
 	return 1;
 }
--- a/fs/coda/file.c~fs-consistently-use-file_has_valid_mmap_hooks-helper
+++ a/fs/coda/file.c
@@ -160,7 +160,7 @@ coda_file_mmap(struct file *coda_file, s
 	size_t count;
 	int ret;
 
-	if (!host_file->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(host_file))
 		return -ENODEV;
 
 	if (WARN_ON(coda_file != vma->vm_file))
--- a/fs/ecryptfs/file.c~fs-consistently-use-file_has_valid_mmap_hooks-helper
+++ a/fs/ecryptfs/file.c
@@ -193,7 +193,7 @@ static int ecryptfs_mmap(struct file *fi
 	 * natively.  If FILESYSTEM_MAX_STACK_DEPTH > 2 or ecryptfs
 	 * allows recursive mounting, this will need to be extended.
 	 */
-	if (!lower_file->f_op->mmap)
+	if (!file_has_valid_mmap_hooks(lower_file))
 		return -ENODEV;
 	return generic_file_mmap(file, vma);
 }
_
